### ISSUE TITLE
Tracking issue for optimizer graph not being an inference graph

### ISSUE THREAD

**Issue by mlazos** [pytorch/pytorch#104053]
### 🐛 Describe the bug

The optimizer is incorrectly classified as a training graph in aot. The main reason this occurs is that some parameters have `requires_grad=True` *and* torchdynamo calls set_grad_enabled(True|False) while simultaneously putting a call_function node in the graph which performs the global state change. 

As a result, when dynamo traces a no_grad context manager, it will trace the *exit* of the context manager *before* calling the user compiler. As a result, by the time we reach aot, grad is enabled, and the graph is treated as a training graph.

Why this matters:
Functionalization appends an epilogue which aggregates all mutation updates of tensors into a single code block consisting of a bunch of copies. In inference mode, these copies are visible to inductor while in training mode, these copies are not visible to inductor. If the copies are not visible to inductor, inductor will allocate tensors for every single parameter in the graph in order for them to get sent to the epilogue, which is atrocious for perf. 

Solutions:
1. graph break before grad is re-enabled
2. in functionalization make the epilogue visible in training mode
3. with compiled backwards, the entire graph is an inference graph by default

1 is two lines of code, so that approach is the easiest. 
3 is getting implemented currently, so once it's ready we can remove the graph breaks.

### Error logs

N/A

### Versions

N/a

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519

**Comment by janeyx99** [comment]
Do we have an expected timeline for (3)? That would be the best solution

**Comment by ezyang** [comment]
We could also potentially make the inference-training detection in AOTAutograd smarter. For example, if no outputs of the graph require_grad (post functionalization), then we do not actually need to compile a training graph, because there is no way to differentiate through it. Post functionalization here is important, because a differentiable mutation to a requires grad input would entail training graph (but a mutation under no_grad would not, as is the case with optimizers.)

**Comment by janeyx99** [comment]
Has there been a solution to this? I was looking through the code and I think the patching that happens for optimizers should be minimized if possible. Today, if someone runs their optimizer through dynamo (so if the dynamo TorchPatcher gets triggered), the optimizer's step will be overridden by an unwrapped, incorrect version of step. The wrapper it gets rid of is the one that tells the optimizer to be in no_grad mode, which is pretty important.

What this means is that once an optimizer has been graced by the TorchPatcher, it _NO LONGER WORKS_ for the eager use case. Maybe this is not a common use case, but the fact that dynamo permanently patches the Optimizer object isn't great for debuggability/understandability from the user perspective.

**Comment by bdhirsh** [comment]
> . For example, if no outputs of the graph require_grad (post functionalization), then we do not actually need to compile a training graph, because there is no way to differentiate through it

We actually do this now! @mlazos can you check to see if any the issues you mentioned are still a problem?

**Comment by mlazos** [comment]
@bdhirsh this is fixed! thanks

**Comment by mlazos** [comment]
Looks like this causes a large performance regression, this will require some investigation

**Comment by janeyx99** [comment]
@mlazos has the following from the issue description been resolved?
> If the copies are not visible to inductor, inductor will allocate tensors for every single parameter in the graph in order for them to get sent to the epilogue, which is atrocious for perf.

**Comment by bdhirsh** [comment]
> If the copies are not visible to inductor, inductor will allocate tensors for every single parameter in the graph in order for them to get sent to the epilogue, which is atrocious for perf.

(fwiw - this should be fixed on the AOTAutograd side)

**Comment by sanchitintel** [comment]
Hi @mlazos, could you please clarify if these graph breaks are solely responsible for eager-mode training being slower than `torch.compile` (on CPU)?

In [profiling data](https://gist.github.com/sanchitintel/c2ccda7bdd58be9c12ecf16fa4680f25) of [a simple workload](https://gist.github.com/sanchitintel/31acb5ff721884b443bdd8a5bd4953b9), despite using `torch._inductor.config.cpp.enable_kernel_profile=True`, I see that `Torch-compiled region` is listed as a bottleneck.

But this profiling data seems to be on a low granularity, and doesn't seem to indicate why `self-time` was high for `Torch-compiled region`s. Is that because all the computation inside `Torch-compiled region`s is not recorded by the profiler, and is thus unaccounted for, as in the empty/gray regions below?
<img width="879" alt="image" src="https://github.com/pytorch/pytorch/assets/90872293/e945de19-adbc-48e0-bcab-e816b9651ff7">

Just trying to ascertain if it's currently possible to minimize this overhead before the graph breaks would be removed in the future (as part of resolving this GitHub issue). Thanks!

**Comment by zou3519** [comment]
@mlazos @bdhirsh Btw we're planning on allowing .backward() calls into the Dynamo-produced graph, under certain situations. The goal is to capture a full training loop (forward-backward-step).

This [graph break](https://github.com/pytorch/pytorch/blob/main/torch/optim/optimizer.py#L65) is in the way. What would it take to remove it?

My cursory reading of this thread is that we need to ensure that AOTDispatcher treats the optimizer graph as an inference graph? (unless that was fixed?)

**Comment by bdhirsh** [comment]
Oh you know what - @mlazos I'm not familiar with that piece of optimizer code, but we did a lot of work to allow input mutations to remain in the graph in as many cases as possible. In particular:

if you have a graph that is not inference-only (some gradients need to be computed), and it also has some input mutations in it (say from an optimizer step), then we **will** keep the mutations in the graph as `copy_` nodes as long as the mutations all happen under `no_grad` in the user code.

I think you can get rid of this graph break?

**Comment by janeyx99** [comment]
@mlazos had tried removing these before but the PR got reverted due to perf regressions -- do we expect these would be fixed 10 months later?

**Comment by mlazos** [comment]
> [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/4105940?s=64&amp;v=4">@mlazos](https://github.com/mlazos?rgh-link-date=2025-02-12T20%3A53%3A15.000Z) had tried removing these before but the PR got reverted due to perf regressions -- do we expect these would be fixed 10 months later? See comment from [#118987 (comment)](https://github.com/pytorch/pytorch/pull/118987#issuecomment-2035883464)

Yes, that's what Brian is talking about. The copy_ nodes will be there now and visible to inductor to fuse.

**Comment by mlazos** [comment]
Yeah we can try removing it, @c-rizz submit a PR and add me as a reviewer. Thanks!

**Comment by c-rizz** [comment]
@mlazos do you mean a PR removing the break altogether or having the break conditionally as in [this commit](https://github.com/pytorch/pytorch/commit/4076c39bd036c0dc6165aca4ae64fd6fa76f3b91)? I can do that either way

**Comment by benjaminglass1** [comment]
Pushed a test branch removing the graph break entirely, and I don't think we're ready to do it. See https://hud.pytorch.org/benchmark/compilers_regression?renderGroupId=main&time.start=2026-06-10T00%3A00%3A00.000Z&time.end=2026-06-17T23%3A59%3A59.999Z&filters.repo=pytorch%2Fpytorch&filters.benchmarkName=compiler&filters.backend=&filters.mode=training&filters.dtype=amp&filters.deviceName=cuda+%28h100%29&filters.device=cuda&filters.arch=h100&lbranch=main&rbranch=gh%2Fbenjaminglass1%2F145%2Forig&rcommit.commit=2bd562526e569ef6850c06fe37fcff6885224453&rcommit.branch=gh%2Fbenjaminglass1%2F145%2Forig&rcommit.date=2026-06-16T23%3A00%3A00Z&rcommit.workflow_id=27652807899&lcommit.commit=703d775a75e4a788fa4df8b4fecb056f28b3a3bc&lcommit.branch=main&lcommit.date=2026-06-17T02%3A00%3A00Z&lcommit.workflow_id=27659575800&maxSampling=110 for test results, but the TL;DR is that most models slowed down when training by removing the break.

**Comment by c-rizz** [comment]
Wouldn't the approach of [this commit](https://github.com/pytorch/pytorch/commit/4076c39bd036c0dc6165aca4ae64fd6fa76f3b91) in [this PR](https://github.com/pytorch/pytorch/pull/173388) at least reduce the impact? In the (common) case in which the grad state does not need to be changed, the break can be skipped and nothing should change
